Add documentation for Kubeflow Trainer v2 TrainJob integration with Kueue

NarayanaSabari opened the pull request:
#### What type of PR is this?

/kind documentation
/sig scheduling
/area batch

#### What this PR does / why we need it:

This PR adds comprehensive documentation for using Kubeflow Trainer v2 TrainJob with Kueue for workload management and scheduling. The documentation covers:

- Overview of TrainJob integration with Kueue
- Queue selection and suspend field configuration
- Using ClusterTrainingRuntime (cluster-scoped) with complete PyTorch DDP example
- Using TrainingRuntime (namespace-scoped) for team-specific custom configurations
- Priority classes with Kueue PriorityClass
- LLM fine-tuning use cases with TorchTune
- Gang scheduling configuration with ClusterQueue
- Monitoring TrainJob status with kubectl
- Troubleshooting common issues (admission, startup, pod creation)
- Best practices for production deployments
- Migration notes highlighting differences from Training Operator v1

The documentation is structured to guide users from basic setup to advanced scenarios, with verified examples that match the actual Trainer v2 API.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #7345 https://github.com/kubeflow/trainer/issues/2919

#### Special notes for your reviewer:

- All API versions, field names, and label formats have been verified against the actual Trainer v2 codebase
- Examples use the correct queue label: `kueue.x-k8s.io/queue-name`
- RuntimeRef structure matches the Go types defined in `pkg/apis/trainer/v1alpha1/trainjob_types.go`
- ClusterTrainingRuntime examples align with manifests in `manifests/base/runtimes/`
- Documentation follows the existing style and structure of other Kubeflow integration docs (TFJob, PyTorchJob, etc.)
- Updated the index page to prominently feature Trainer v2 as the recommended option

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

mimowo commented on site/content/en/docs/tasks/run/kubeflow/trainjobs.md:
````diff
@@ -0,0 +1,374 @@
+        memory: "32Gi"
+        nvidia.com/gpu: "2"
+```
+
````
This section seems very generic for all Job types supported in queue, drop unless we have something specific

andreyvelich replied:
Yes, probably, we can just at the beginning link to the BuiltinTrainer guides: https://www.kubeflow.org/docs/components/trainer/user-guides/builtin-trainer/torchtune/

mimowo commented on site/content/en/docs/tasks/run/kubeflow/trainjobs.md:
````diff
@@ -0,0 +1,374 @@
+        nvidia.com/gpu: "1"
+```
+
+## LLM Fine-Tuning with Kueue
````
is there such an example I. the kubełet docs? if so we could refere there and mention that the only difference is the added queue name label

NarayanaSabari commented on site/content/en/docs/tasks/run/kubeflow/trainjobs.md:
```diff
@@ -0,0 +1,374 @@
+metadata:
+  name: high-priority-training
+  namespace: default
+  labels:
```
Agreed! I'll update this section to reference the workload priority class documentation at `/docs/tasks/run/jobs_with_priority` instead of providing a full example. I'll keep just a brief note showing that TrainJobs use the same `kueue.x-k8s.io/priority-class` label as other workload types.

andreyvelich commented on site/content/en/docs/tasks/run/kubeflow/trainjobs.md:
```diff
@@ -0,0 +1,258 @@
+    image: docker.io/kubeflow/pytorch-dist-mnist-test:v1.0
+    command:
+      - torchrun
+      - /workspace/examples/mnist/mnist.py
```
Please can you verify that this image is still functional ?

NarayanaSabari replied:
Good catch! I need to verify that `docker.io/kubeflow/pytorch-dist-mnist-test:v1.0` still works. 

Could you recommend a replacement if this image is deprecated? I noticed the Trainer repo has several training images - should I use one from the current examples like:
- `docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727` 
- Or another image from the Trainer manifests?

I'll test and update accordingly.

andreyvelich replied:
IMO, you can just use the default PyTorch image from the `torch-distributed` runtime in your examples: https://github.com/kubeflow/trainer/blob/master/manifests/base/runtimes/torch_distributed.yaml#L25
Since this image is executable, you can create TrainJob as this:
```
apiVersion: trainer.kubeflow.org/v1alpha1
kind: TrainJob
metadata:
  name: pytorch-distributed
  labels:
    kueue.x-k8s.io/queue-name: user-queue
spec:
  runtimeRef:
    name: torch-distributed
    kind: ClusterTrainingRuntime
```

In the future, we can add more examples which shows how to use Kubeflow SDK, TrainJob, and Kueue.

NarayanaSabari commented on site/content/en/docs/tasks/run/kubeflow/trainjobs.md:
```diff
@@ -0,0 +1,258 @@
+- **ClusterTrainingRuntime**: Cluster-scoped training runtimes that can be used across all namespaces
+- **TrainingRuntime**: Namespace-scoped training runtimes that are only available within a specific namespace
+
+Kueue manages TrainJobs by scheduling their underlying JobSets according to available quota and priority.
```
✅ Fixed! Changed "underlying JobSets" to "underlying jobs" to account for future Slurm/Flux support. Thanks for catching this!

NarayanaSabari commented:
@andreyvelich Thank you for the comprehensive review! I've addressed all your comments:

**Summary of Changes:**
- ✅ Fixed "JobSet" → "jobs" wording  
- ✅ Updated installation link to operator guide
- ✅ Added ClusterTrainingRuntime guide cross-reference
- ✅ Updated LLM section to reference BuiltinTrainer guides
- ✅ Enhanced Additional Resources section

**Pending:**
- 🔍 Docker image verification - waiting for your recommendation
- 💬 Website restructuring - need guidance on preferred approach

Also addressed all of @mimowo's feedback (removed generic sections, added references to existing Kueue docs).

Thanks for helping improve this documentation!

mimowo commented on site/content/en/docs/tasks/run/kubeflow/trainjobs.md:
```diff
@@ -0,0 +1,261 @@
+date: 2024-11-05
+weight: 7
+description: >
+  Run a Kueue scheduled TrainJob from Kubeflow Trainer v2
```
Seems like some rendering issue to solve:

<img width="1489" height="734" alt="Image" src="https://github.com/user-attachments/assets/2df3f552-720b-44fc-9898-4e936e344906" />

NarayanaSabari commented on site/content/en/docs/tasks/run/kubeflow/trainjobs.md:
```diff
@@ -0,0 +1,258 @@
+---
```
@mimowo @andreyvelich 

Before implementing the website restructuring, I need your approval on the approach:

## Current File Structure
```
docs/tasks/run/kubeflow/
├── _index.md (title: "Kubeflow Jobs")
├── trainjobs.md (NEW - TrainJob v2)
├── mpijobs.md
├── pytorchjobs.md
├── tfjobs.md
├── xgboostjobs.md
├── paddlejobs.md
└── jaxjobs.md
```

## Proposed Approach

**Keep current file locations**, update only `_index.md` to show this navigation tree:

```
Kubeflow Training
├── Trainer v2
│   └── Run a TrainJob
├── Trainer v1
│   ├── Run a PyTorchJob
│   ├── Run a TFJob
│   ├── Run an XGBoostJob
│   ├── Run a PaddleJob
│   └── Run a JAXJob
└── MPI Operator
    └── Run an MPIJob
```

**Implementation:** Only modify `_index.md` content, keep all files in same location.

## Benefits
- ✅ No file moves (existing links keep working)
- ✅ Clear v2 vs v1 distinction
- ✅ Easy to update when v2 goes stable

## Questions
1. Is this structure acceptable?
2. Should I implement in **this PR** or **follow-up PR**?

Please confirm and I'll proceed. Thanks!

mimowo replied:
Proposed Approach sgtm

andreyvelich replied:
I suggested to move Kubeflow TrainJob (v2) to the root:
```
# Name this page Kubeflow TrainJob (v2)
/docs/tasks/run/trainjob/

# Rename the below section to Kubeflow Jobs (v1)
/docs/tasks/run/kubeflow/jaxjobs
/docs/tasks/run/kubeflow/tfjob


```

NarayanaSabari replied:
@andreyvelich @mimowo Updated are made, good to go

andreyvelich replied:
@NarayanaSabari We don't really need to have subsection Kubeflow TrainJob (v2), since it only has one page.
Can we create similar page as for RayJobs without nested structure: https://github.com/kubernetes-sigs/kueue/blob/main/site/content/en/docs/tasks/run/rayjobs.md ?

Also, please can you update this page: https://kueue.sigs.k8s.io/docs/tasks/#batch-user ?

andreyvelich commented:
@NarayanaSabari We might also want to introduce Kubeflow TrainJob page in the multi-cluster: https://deploy-preview-7533--kubernetes-sigs-kueue.netlify.app/docs/tasks/run/multikueue/
Since we support MultiKueue in TrainJob.
